lib/logstash/pluginmanager/util.rb

```diff
@@ -1,64 +1,101 @@
+        $stderr.puts(e.message) if ENV["DEBUG"]
+        return false
+      end
+    else
+      dep = Gem::Dependency.new(plugin, version || Gem::Requirement.default)
+      specs, error = Gem::SpecFetcher.fetcher.spec_for_dependency(dep)
 
-    gem_data = case
-    when gem.is_a?(Gem::Specification); gem
-    when (gem =~ /\.gem$/ and File.file?(gem)); Gem::Package.new(gem).spec
-    else Gem::Specification.find_by_name(gem)
+      # depending on version requirements, multiple specs can be returned in which case
+      # we will grab the one with the highest version number
+      if latest = specs.map{|tuple| tuple.first}.max_by{|spec| spec.version}
+        unless valid = is_logstash_plugin_gem_spec?(latest)
+          $stderr.puts("#{plugin} is not a Logstash plugin")
+        end
+        return valid
+      else
+        $stderr.puts("Plugin #{plugin}" + (version ? " version #{version}" : "") + " does not exists")
+        return false
+      end
     end
+  end
 
-    gem_data.metadata['logstash_plugin'] == "true" ? gem_data : false
+  # @param spec [Gem::Specification] plugin gem specification
+  # @return [Boolean] true if this spec is for an installable logstash plugin
+  def self.is_logstash_plugin_gem_spec?(spec)
+    spec.metadata["logstash_plugin"] == "true"
   end
 
-  def self.download_gem(gem_name, gem_version = '')
- 
-    gem_version ||= Gem::Requirement.default
- 
-    dep = ::Gem::Dependency.new(gem_name, gem_version)
-    specs_and_sources, errors = ::Gem::SpecFetcher.fetcher.spec_for_dependency dep
-    if specs_and_sources.empty?
-      return false
-    end
-    spec, source = specs_and_sources.max_by { |s,| s.version }
-    path = source.download( spec, java.lang.System.getProperty("java.io.tmpdir"))
-    path
+  # @param path [String] path to .gem file
+  # @return [Gem::Specification] .get file gem specification
+  # @raise [Exception] Gem::Package::FormatError will be raised on invalid .gem file format, might be other exceptions too
+  def self.plugin_file_spec(path)
+    Gem::Package.new(path).spec
   end
 
-  def self.installed?(name)
-    Gem::Specification.any? { |x| x.name == name }
+  # @param plugin [String] the plugin name or the local path to a .gem file
+  # @return [Boolean] true if the plugin is a local .gem file
+  def self.is_plugin_file?(plugin)
+    (plugin =~ /\.gem$/ && File.file?(plugin))
   end
 
-  def self.matching_specs(name)
-    req = Gem::Requirement.default
-    re = name ? /#{name}/i : //
-    specs = Gem::Specification.find_all{|spec| spec.name =~ re && req =~ spec.version}
-    specs.inject({}){|result, spec| result[spec.name_tuple] = spec; result}.values
+  # retrieve gem specs for all or specified name valid logstash plugins locally installed
+  # @param name [String] specific plugin name to find or nil for all plungins
+  # @return [Array<Gem::Specification>] all local logstash plugin gem specs
+  def self.find_plugins_gem_specs(name = nil)
```

Comment:
ultimately its all plural results but in practice `find_all_by_name` should only return a 1 element array... so...